encoding/wave/decode.go

```diff
@@ -0,0 +1,217 @@
+	"encoding/binary"
+	"fmt"
+	"io"
+	"reflect"
+)
+
+type (
+	// Decoder is a WAVE decoder
+	Decoder struct {
+		input io.Reader
+	}
+)
+
+// NewDecoder creates a new WAVE decoder
+func NewDecoder(r io.Reader) *Decoder {
+	return &Decoder{
+		input: r,
+	}
+}
+
+// Decode the WAVE into data array and returns the WAVE header
+func (d *Decoder) Decode(data interface{}) (Header, error) {
+	v := reflect.ValueOf(data)
+	if v.Kind() != reflect.Ptr || v.IsNil() {
+		return Header{}, fmt.Errorf("Decode expects a non-nil pointer")
+	}
+	deref := v.Elem()
+	switch deref.Kind() {
+	case reflect.Slice:
+		// get the underlying type of slice elements
+		switch typ := deref.Type().Elem().Kind(); typ {
+		case reflect.Int16, reflect.Float32:
+			return d.decode(v, typ)
+		default:
+			return Header{}, fmt.Errorf("unknown slice type: %v",
+				deref.Type().Elem().Kind())
+		}
+	}
+
+	return Header{}, fmt.Errorf("impossible to decode wave into %s. "+
+		"Expect *[]int16, *[]float32", v.Kind())
+}
+
+func (d *Decoder) decode(v reflect.Value, typ reflect.Kind) (Header, error) {
+	hdr, err := d.parseHeader()
+	if err != nil {
+		return Header{}, err
+	}
+
+	switch typ {
+	case reflect.Int16:
+		if hdr.RiffChunkFmt.AudioFormat != FormatPCM {
+			return hdr, fmt.Errorf("[]int16 requires a PCM audio")
+		}
+		err = d.decodeInt16(v, hdr.DataBlockSize)
+	case reflect.Float32:
+		if hdr.RiffChunkFmt.AudioFormat != FormatIEEEFloat {
+			return hdr, fmt.Errorf("[]float32 requires a IEEEFloat format")
+		}
+		err = d.decodeFloat32(v, hdr.DataBlockSize)
+	}
+	return hdr, err
+}
+
+func (d *Decoder) decodeInt16(v reflect.Value, datasz uint32) error {
+	dataptr := v.Interface().(*[]int16)
+
+	const typesize = 2
+	var bytesRead uint32
+	for bytesRead < datasz {
+		var buf [typesize]byte
+		n, err := d.input.Read(buf[:])
+		if err != nil {
+			return err
+		}
+		if n != typesize {
+			return fmt.Errorf("corrupted audio")
+		}
+
+		sample := int16(binary.LittleEndian.Uint16(buf[:]))
+		*dataptr = append(*dataptr, sample)
+		bytesRead += typesize
+	}
+
+	return nil
+}
+
+func (d *Decoder) decodeFloat32(v reflect.Value, datasz uint32) error {
+	const maxval float32 = 1.0
+	const minval float32 = -1.0
+	const typesz = 4
+
+	dataptr := v.Interface().(*[]float32)
+	var bytesRead uint32
+	for bytesRead < datasz {
+		var sample float32
+		err := binary.Read(d.input, binary.LittleEndian, &sample)
+		if err != nil {
+			return fmt.Errorf("decoding floats: bytes read[%d], total[%d]: %s",
+				bytesRead,
+				datasz,
+				err)
+		}
+
+		if sample < minval || sample > maxval {
+			return fmt.Errorf(
+				"sample[%f] is outside the valid value range for a PCM float",
+				sample,
+			)
+		}
+
+		*dataptr = append(*dataptr, sample)
+		bytesRead += typesz
+	}
+
+	return nil
+}
+
+func (d *Decoder) parseRIFFHdr() (RiffHeader, error) {
+	var hdr RiffHeader
+	err := binary.Read(d.input, binary.LittleEndian, &hdr)
+	if err != nil {
+		return RiffHeader{}, fmt.Errorf("parsing riff: %s", err)
```

Comment:
Remember when we discussed using pointers to use nil here ? The case where I got confused is when we have strings and errors on the return, for example:

```go
func something() (string, error) {
    return "", errors.New("etc")
}
```

If initializing an empty object is terrible we would have to use string* too, but this feels clumsy =P.
But I also dislike initializing empty structs...so I feel your feelings =) (not sure if you still have them)

Reply:
be careful with purisms =) The worse sometimes is better hehe

I dislike having to initialize the empty value of a long object also, but little-and-concise names to the rescue =)

Reply:
Indeed =)